packages/@aws-cdk/aws-appsync/lib/schema.ts

```diff
@@ -59,9 +78,16 @@ export class Schema {
   private types: IIntermediateType[];
 
   public constructor(options?: SchemaOptions) {
+    if (options?.filePath && options?.s3Location) {
```

Comment:
The fact that you have to write this validation suggests to me this API could be better.

Can we actually change this API here a little bit?

1. Make `Schema` abstract.
2. Create a subclass of `Schema`, `SchemaImpl`, that's module-private, and contains all of the logic here. It extends `Schema`, and takes `SchemaOptions` in its constructor argument. `SchemaOptions` should now also be module-private.
3. All of the static factory methods declare that they return `Schema` in their signature, but return `SchemaImpl` in their bodies.

This way, you will be in control of the invariants of `SchemaImpl`, and you won't have to write this defensive code. In fact, you can have a separate subclass of `Schema` for each of the static factory methods!

Reply:
I'm all for this change, but it would be a breaking change because we might have users using the `Schema` api using `new Schema()`. Are we comfortable with introducing this breaking change?

Reply:
Yep, [`appsync` is still experimental](https://github.com/aws/aws-cdk/blob/1eb6de7913a27874a2c8c382900d6976fe7f0564/packages/%40aws-cdk/aws-appsync/package.json#L124-L125).

Reply:
Sweet, ill get it done today